### PR TITLE
Add sd_notify for systemd start checking

### PR DESCRIPTION
**Pull request by bigbes**
Rewritten sd_notify -> `systemd_notify`/`snotify`/`vsnotify`, e.t.c.

Proposal -> add `cfg_gets` (values `loaded`, `before loaded`, `custom`) for systemd for user to send `READY` from Lua when it wants.

### CODE REVIEW

File: src/systemd.c [hunk 1]
```diff
@@ -0,0 +1,127 @@
...
+		.msg_name = &sa,
+	};
+
+	strncpy(sa.sun_path, sd_unix_path, sizeof(sa.sun_path));
+	if (sa.sun_path[0] == '@')
+		sa.sun_path[0] = '\0';
+
+	msg.msg_namelen = sizeof(sa.sun_family) + strlen(sa.sun_path);
+
+	ssize_t sent = sendmsg(systemd_fd, &msg, MSG_NOSIGNAL);
```
Comment (rtsisyk) [inline src/systemd.c]: I noted that systemd uses special [cmsg types](https://github.com/systemd/systemd/blob/f6f372d2f46ac9be7cbb1ecd8f82f1b3ab669924/src/libsystemd/sd-daemon/sd-daemon.c#L543-L553) for messages with pids. Probably we need to implement systemd_notify and systemd_notify_with_pid() to support this business logic.

**Comment by rtsisyk** [comment]
> Proposal -> add cfg_gets (values loaded, before loaded, custom) for systemd for user to send READY from Lua when it wants.

Sounds interesting, but raises a lot of questions. Let's finish sd_notify feature first and discuss this proposal later.

File: src/systemd.c (hunk 1 above)
Comment (bigbes) [inline src/systemd.c]: No need to do this, in case if pid==0. It's exactly this case

File: src/systemd.c [hunk 2]
```diff
@@ -0,0 +1,127 @@
...
+	free(buf);
+	return rv;
+}
+
+/*
+ * @return  0 on non-systemd platforms
+ * @return -1 on errors (more information in errno)
+ * @return >0 on ok
+ */
+__attribute__ ((format (printf, 1, 2))) int
```
Comment (bigbes) [inline src/systemd.c]: ok, sorry, will fix

File: src/main.cc [hunk 3]
```diff
@@ -513,6 +518,7 @@ tarantool_free(void)
 	memory_free();
 	random_free();
 #endif
+	systemd_shutdown();
```
Comment (bigbes) [inline src/main.cc]: I've guessed that because antonym for `init` is `shutdown`. `free` is antonym for `allocate`

File: src/systemd.h [hunk 4]
```diff
@@ -0,0 +1,49 @@
...
+ */
+
+#if defined(__cplusplus)
+extern "C" {
+#endif /* defined(__cplusplus) */
+
+int systemd_init();
+void systemd_shutdown();
+
+int systemd_notify(const char *state);
```
Comment (bigbes) [inline src/systemd.h]: Move them from `.c` to `.h` file?

**Comment by bigbes** [comment]
Everything is done

**Comment by kostja** [comment]
* Eugine Blikh <notifications@github.com> [17/05/03 19:57]:
> Everything is done

How does it work on BSD*?

-- 
Konstantin Osipov, Moscow, Russia, +7 903 626 22 32
http://tarantool.org - www.twitter.com/kostja_osipov

**Comment by bigbes** [comment]
@kostja what is different on *BSD?

**Comment by bigbes** [comment]
Oh, I've investigated: no systemd - no problems.

File: src/systemd.c [hunk 5]
```diff
@@ -0,0 +1,127 @@
+#include <errno.h>
```
Comment (bigbes) [inline src/systemd.c]: done

**Pull request merged**